Make the members of `KeyPair` the same as the members of `Pair`

linked0 opened the pull request:
We should merge the `KeyPair` and the `Pair` types into one. I made the components of the classes the same in this PR as the middle stage.
`SecretKey` of the `KeyPair` and `Scalar` of the `Pair` have `BitBlob!(crypto_core_ed25519_SCALARBYTES * 8)` data in this PR. And `PublicKey` of the `KeyPair` and `Point` of the `Pair` have `BitBlob!(crypto_core_ed25519_BYTES * 8)` data.

Feel free to post any opinions and suggestions. I will have a whiteboard if needed.

Fixes #1645

linked0 committed:
Add code checking `key_pair`

In the constructor of `EnrollmentManager`, the `key_pair` parameter could
be `KeyPair.init` because the node is not validator. So we should not
convert the parameter to the `Pair` type.

linked0 committed:
Make the `KeyPair` same as the `Pair`

In order to merge the `KeyPair` and `Pair`, we should make the components of
the classes have same data type. So we made the `PublicKey` the type of
`Point` and make the `SecretKey` the type of `Scalar`.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -64,10 +66,15 @@ public struct KeyPair
     {
-        SecretKey sk;
+        BitBlob!(crypto_sign_ed25519_SECRETKEYBYTES * 8) sk_data;
         PublicKey pk;
-        if (crypto_sign_seed_keypair(pk[].ptr, sk[].ptr, seed[].ptr) != 0)
+        if (crypto_sign_seed_keypair(pk[].ptr, sk_data[].ptr, seed[].ptr) != 0)
             assert(0);
+
+        Scalar x25519_sk;
+        if (crypto_sign_ed25519_sk_to_curve25519(cast(ubyte*)(x25519_sk[].ptr), sk_data[].ptr) != 0)
+            assert(0);
+        SecretKey sk = SecretKey(x25519_sk[]);
```
Are those different ? I expect they just hold the same value, don't they ?

linked0 replied:
No, they are not. The data of the `SecretKey` has the `seed` and `Point` in its 64 bytes storage as follows.
https://github.com/jedisct1/libsodium/blob/57d950a54e6f7743084092ba5d31b8fa0641eab2/src/libsodium/crypto_sign/ed25519/ref10/keypair.c#L26-L27

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -248,9 +260,13 @@ public struct PublicKey
-        return 0 ==
-            crypto_sign_ed25519_verify_detached(
-                signature[].ptr, msg.ptr, msg.length, this.data[].ptr);
+        // return 0 ==
+        //     crypto_sign_ed25519_verify_detached(
+        //         signature[].ptr, msg.ptr, msg.length, this.data[].ptr);
+
+        import agora.crypto.Schnorr;
+        auto X = Point(this.data[]);
+        return agora.crypto.Schnorr.verify(X, signature, msg);
```
Don't comment out code, remove it. Also if we're switching to `verify` then we don't need the `DataType` aliases anymore, do we ?

linked0 replied:
Removed the comment.

But I don't understand why this change is related to the `DataType` alias. Anyway did you mean this kind of change?

```
public struct PublicKey
{
    BitBlob!(crypto_core_ed25519_BYTES * 8) data;
    alias data this;

    /// Construct an instance from binary data
    public this (const BitBlob!(crypto_core_ed25519_BYTES * 8) args)
        pure nothrow @safe @nogc
    {
        this.data = args;
    }
```

hewison-chris reviewed (approved):
LGTM, but I think in the commit description you have the Point and Scalar swapped.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -123,7 +132,7 @@ public struct PublicKey
     public string toString () const @trusted nothrow
     {
-        ubyte[1 + PublicKey.Width + 2] bin;
+        ubyte[1 + crypto_core_ed25519_BYTES + 2] bin;
```
Revert, because they should be the same

Geod24 replied:
I see why you need it now. We can just use `sizeof` here.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -133,7 +142,7 @@ public struct PublicKey
     public void toString (scope void delegate(const(char)[]) sink) const @trusted
     {
-        ubyte[1 + PublicKey.Width + 2] bin;
+        ubyte[1 + crypto_core_ed25519_BYTES + 2] bin;
```
Revert

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -169,7 +178,7 @@ public struct PublicKey
     {
         const bin = Base32.decode(str);
-        enforce(bin.length == 1 + PublicKey.Width + 2);
+        enforce(bin.length == 1 + crypto_core_ed25519_BYTES + 2);
```
Revert

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -245,12 +254,7 @@ public struct PublicKey
-        return 0 ==
-            crypto_sign_ed25519_verify_detached(
-                signature[].ptr, msg.ptr, msg.length, this.data[].ptr);
+        return agora.crypto.Schnorr.verify(Point(this.data[]), signature, msg);
```
Just `this.data` ? It's already a `Point`

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -348,12 +348,13 @@ public struct SecretKey
-        if (crypto_sign_ed25519_detached(result[].ptr, null, msg.ptr, msg.length, this.data[].ptr) != 0)
-            assert(0);
-        return result;
+        Pair kp = Pair.fromScalar(Scalar(this.data[]));
```
I think we should drop `SecretKey` completely and just move this to `KeyPair`.

Geod24 replied:
Also, `this.data` is already a `Scalar`

Geod24 replied:
Let's keep removing `SecretKey` for later.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -348,12 +348,13 @@ public struct SecretKey
-            assert(0);
-        return result;
+        Pair kp = Pair.fromScalar(Scalar(this.data[]));
+        version (unittest)
+        {
+            const Pair noise = Pair.fromScalar(Scalar(hashMulti(kp.v, "sign")));
+            return agora.crypto.Schnorr.sign(kp, noise, msg);
+        }
```
What is this ?

linked0 replied:
This is for the unit test of `PrettyPrinter` checking the formatted output of a created block. A random noise for signing changes every time as you know. Is there any better way for that?

hewison-chris replied:
I will take a look.

linked0 replied:
Deleted with the support of @hewison-chris.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -540,11 +541,7 @@ unittest
-    if (crypto_sign_ed25519_sk_to_curve25519(cast(ubyte*)(x25519_sk[].ptr), secret[].ptr) != 0)
-        assert(0);
-    return x25519_sk;
+    return Scalar(secret[]);
```
`return secret.data` ?

Geod24 commented on source/agora/consensus/Quorum.d:
```diff
@@ -115,7 +115,10 @@ public QuorumConfig buildQuorumConfig (in PublicKey key,
     }
 
-    quorum.nodes.sort;
+    () @trusted {
+        scope (failure) assert(0);
+        quorum.nodes.sort;
+    }();
```
This looks unrelated ?

Geod24 replied:
I see why. You get the error message because, while sorting, `opCmp` is called, and unfortunately it is not annotated. Will fix.

Geod24 replied:
For reference:
https://github.com/bpfkorea/crypto/blob/374d775327cb381a5037deb88edb7e352636de2f/source/agora/crypto/ECC.d#L382-L392
Calls https://github.com/Geod24/bitblob/blob/5418160fe7751eaa1ec99870ae7c62bd1f9c58e7/source/geod24/bitblob.d#L237-L251 ~but the later doesn't require annotations because it's templated~ Edit: Nvm the attributes are at the top.

Geod24 replied:
Depends on https://github.com/bpfkorea/crypto/pull/19

hewison-chris replied:
BTW crypto submodule update has been merged.
Commit cbb4169a3a23092d18463b8723a68680f36c5010  includes the changes to attributes.

hewison-chris replied:
Will be included now if you rebase on upstream.

linked0 replied:
OK, I will apply it. Thanks.

linked0 replied:
Applied new `crypto` submodule.

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -565,7 +562,7 @@ unittest
     Signature enroll_sig = sign(pair, "BOSAGORA");
 
-    Point point_Address = Point(kp.address);
+    Point point_Address = Point(kp.address[]);
```
I think we can just remove this unittest now ?

Geod24 commented on source/agora/consensus/validation/Enrollment.d:
```diff
@@ -66,7 +66,7 @@ public string isInvalidReason (const ref Enrollment enrollment,
         return "Enrollment: UTXO is not frozen";
 
-    Point address = Point(utxo_set_value.output.address);
+    Point address = Point(utxo_set_value.output.address[]);
```
Why do we need to slice ? `address` has an `alias this` to a `Point`.

Geod24 reviewed (approved):
Broadly LGTM

Geod24 commented:
Pushed some fixes

linked0 commented:
@hewison-chris @Geod24 Thank you for your review and comments. All the suggestions except [one](https://github.com/bpfkorea/agora/pull/1702#discussion_r589043952) are applied. 
I will remove `secretKeyToCurveScalar` and request reviews again.

linked0 committed:
Remove `secretKeyToCurveScalar` function

This function is not needed anymore because the `SecretKey` has the `Scalar`
data now.

linked0 commented:
Removed the `secretKeyToCurveScalar` function which is not needed anymore because the `SecretKey` has the `Scalar` data now.

hewison-chris commented on source/agora/common/crypto/Key.d:
```diff
@@ -102,14 +114,11 @@ unittest
-    private alias DataType = BitBlob!(crypto_sign_ed25519_PUBLICKEYBYTES * 8);
-
-    /*private*/ DataType data;
+    /*private*/ Point data;
```
The commit message is wrong. 
Public key is Point and Secret key is Scalar.

linked0 replied:
Updated. Ty.

hewison-chris commented on source/agora/common/crypto/Key.d:
```diff
@@ -123,7 +132,7 @@ public struct PublicKey
     public string toString () const @trusted nothrow
     {
-        ubyte[1 + PublicKey.Width + 2] bin;
+        ubyte[1 + PublicKey.sizeof + 2] bin;
```
Perhaps descriptive constants could be used for these extra byte lengths.

linked0 replied:
I see what you mean. I will add them in another commit.

linked0 committed:
Descriptive constants for extra byte lengths

In `toString` functions of `PublicKey` and `Seed`, the values for extra
byte lengths are not descriptive. So we made the code more readable by
adopting descriptive constants for them.

hewison-chris reviewed (approved):
LGTM

Geod24 commented on source/agora/common/crypto/Key.d:
```diff
@@ -81,14 +88,19 @@ public struct KeyPair
             assert(0);
-        return KeyPair(pk, sk, seed);
+
+        Scalar x25519_sk;
+        if (crypto_sign_ed25519_sk_to_curve25519(cast(ubyte*)(x25519_sk[].ptr), sk_data[].ptr) != 0)
+            assert(0);
+        SecretKey sk = SecretKey(x25519_sk[]);
+        return KeyPair(PublicKey(pk[]), sk, seed);
```
https://github.com/jedisct1/libsodium/blob/ae4add868124a32d4e54da10f9cd99240aecc0aa/src/libsodium/crypto_sign/ed25519/ref10/keypair.c#L69-L83

For reference
